server/src/main/java/org/jboss/as/server/deployment/annotation/AnnotationIndexSupport.java

```diff
@@ -0,0 +1,170 @@
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.jboss.as.server.deployment.DeploymentUnitProcessingException;
+import org.jboss.as.server.logging.ServerLogger;
+import org.jboss.as.server.moduleservice.ModuleIndexBuilder;
+import org.jboss.jandex.Index;
+import org.jboss.jandex.Indexer;
+import org.jboss.modules.Module;
+import org.jboss.modules.ModuleLoadException;
+import org.jboss.modules.ModuleLoader;
+import org.jboss.modules.Resource;
+import org.jboss.modules.filter.PathFilter;
+import org.jboss.modules.filter.PathFilters;
+import org.wildfly.common.Assert;
+
+/**
+ * Utility class used by {@link CompositeIndexProcessor} to assist with
+ * processing annotation indices for a deployment.
+ */
+public final class AnnotationIndexSupport {
+
+    private final ConcurrentMap<String, Lock> indexLocks = new ConcurrentHashMap<>();
+    private final WeakReference<Map<String, Set<Index>>> indexCache;
+
+    public AnnotationIndexSupport(Map<String, Set<Index>> indexCache) {
+        // Store the cache in a weak reference. The OperationContext whose operations resulted in creation
+        // of this object will hold a strong ref. Once that OperationContext is gc'd this weak ref
+        // can be collected, preventing holding the possibly large indices in memory after completion
+        // of the related deployment operations.
+        this.indexCache = new WeakReference<>(indexCache);
+    }
+
+    /**
+     * Gets the annotation indices for the resources in the module with the given id.
+     * @param moduleId the id of the module to be indexed. Cannot be {@code null}.
+     * @param moduleLoader {@link ModuleLoader} to use to load the module. Cannot be {@code null}.
+     * @return the indices. Will not return {@code null}.
+     * @throws DeploymentUnitProcessingException if a problem occurs obtaining the indices
+     */
+    Set<Index> getAnnotationIndices(String moduleId, ModuleLoader moduleLoader) throws DeploymentUnitProcessingException {
+        Assert.checkNotNullParam("moduleId", moduleId);
+        Assert.checkNotNullParam("moduleLoader", moduleLoader);
+
+        Set<Index> result;
+
+        // See if we have this one in the cache
+        Map<String, Set<Index>> cache = getIndexCache(moduleId);
+        result = cache.get(moduleId);
+        if (result == null) {
+            // Nope. Only one thread at a time can create the index for this module id
+            Lock lock = getModuleIndexLock(moduleId);
+            try {
+                lock.lockInterruptibly();
+                try {
+                    // Check in case they were cached while we waited for the lock
+                    result = cache.get(moduleId);
+                    if (result == null) {
+                        // Nope. We build and cache the indices
+                        result = indexModule(moduleId, moduleLoader);
+                        cache.put(moduleId, result);
```

Comment:
AFAICS this is a HashMap, so the 'per key' lock is not enough, you need to synchronize on all operations, or change it to a CHM.

Reply:
oh, sure, of course. Thanks!